langchain/src/agents/agent_iterator.ts

```diff
@@ -0,0 +1,276 @@
+import {
+  CallbackManager,
+  CallbackManagerForChainRun,
+  Callbacks,
+} from "../callbacks/manager.js";
+import { Serializable } from "../load/serializable.js";
+import { AgentFinish, AgentStep } from "../schema/index.js";
+import { Tool } from "../tools/base.js";
+import type { AgentExecutor } from "./executor.js";
+
+interface AgentExecutorIteratorInput {
+  agentExecutor: AgentExecutor;
+  inputs: Record<string, string>;
+  callbacks?: Callbacks;
+  tags?: string[];
+  metadata?: Record<string, unknown>;
+  runName?: string;
+  runManager?: CallbackManagerForChainRun;
+}
+
+export class AgentExecutorIterator
+  extends Serializable
+  implements AgentExecutorIteratorInput
+{
+  lc_namespace = ["langchain", "agents", "executor", "iterator"];
+
+  agentExecutor: AgentExecutor;
+
+  inputs: Record<string, string>;
+
+  callbacks: Callbacks;
+
+  tags: string[] | undefined;
+
+  metadata: Record<string, unknown> | undefined;
+
+  runName: string | undefined;
+
+  private _finalOutputs: Record<string, unknown> | undefined;
+
+  get finalOutputs(): Record<string, unknown> | undefined {
+    return this._finalOutputs;
+  }
+
+  /** Intended to be used as a setter method, needs to be async. */
+  async setFinalOutputs(value: Record<string, unknown> | undefined) {
+    this._finalOutputs = undefined;
+    if (value) {
+      const preparedOutputs: Record<string, unknown> =
+        await this.agentExecutor.prepOutputs(this.inputs, value, true);
+      this._finalOutputs = preparedOutputs;
+    }
+  }
+
+  runManager: CallbackManagerForChainRun | undefined;
+
+  intermediateSteps: AgentStep[] = [];
+
+  iterations = 0;
+
+  get nameToToolMap(): Record<string, Tool> {
+    const toolMap = this.agentExecutor.tools.map((tool) => ({
+      [tool.name]: tool,
+    }));
+    return Object.assign({}, ...toolMap);
+  }
+
+  constructor(fields: AgentExecutorIteratorInput) {
+    super();
+    this.agentExecutor = fields.agentExecutor;
+    this.inputs = fields.inputs;
+    this.tags = fields.tags;
+    this.metadata = fields.metadata;
+    this.runName = fields.runName;
+    this.runManager = fields.runManager;
+  }
+
+  /**
+   * Reset the iterator to its initial state, clearing intermediate steps,
+   * iterations, and time elapsed.
+   */
+  reset(): void {
+    this.intermediateSteps = [];
+    this.iterations = 0;
+    this._finalOutputs = undefined;
+  }
+
+  /**
+   * Increment the number of iterations and update the time elapsed.
+   */
+  updateIterations(): void {
+    this.iterations += 1;
+  }
+
+  /** Method to initialize the iterator */
+  async *[Symbol.asyncIterator]() {
```

Comment:
We sure we need this? Might be nicer to have a method that returns an `IterableReadableStream`?

Reply:
This is cleaner imo. But, one could make the case that calling a method you invoke would provide a better DX